lib/jxl/enc_ac_strategy.cc

```diff
@@ -580,196 +520,81 @@ void ProcessRectACS(PassesEncoderState* JXL_RESTRICT enc_state,
-        prev = ac_strategy->ConstRow(by + iy)[bx + ix].RawStrategy();
-        MaybeReplaceACS(bx + ix, by + iy, config, cmap_factors, ac_strategy,
-                        entropy_adjust, entropy_estimate + (iy * 8 + ix), block,
-                        scratch_space, quantized);
-      }
-      AcStrategy acs = ac_strategy->ConstRow(by + iy)[bx + ix];
-      for (size_t y = 0; y < acs.covered_blocks_y(); y++) {
-        for (size_t x = 0; x < acs.covered_blocks_x(); x++) {
-          computed_mask[(iy + y) * 8 + ix + x] = 1;
-        }
+      float entropy = 0.0;
+      const uint8_t best_of_8x8s =
+	FindBest8x8Transform(8 * (bx + ix), 8 * (by + iy), config, cmap_factors, 
+			     ac_strategy, entropy_adjust, 
+			     block, 
+			     scratch_space, quantized, &entropy);
+      ac_strategy->Set(bx + ix, by + iy, static_cast<AcStrategy::Type>(best_of_8x8s));
+      entropy_estimate[iy * 8 + ix] = entropy;
+    }
+  }
+  // Attempt to merge if a larger transform is better.
+    /*
+  set(AcStrategy::Type::DCT16X16, 0.0f, 0.83f);
+  set(AcStrategy::Type::DCT64X64, 0.0f, 1.3f);
+  set(AcStrategy::Type::DCT64X32, 0.0f, 1.15f);
+  set(AcStrategy::Type::DCT32X64, 0.0f, 1.15f);
+  set(AcStrategy::Type::DCT32X32, 0.0f, 0.97f);
+  set(AcStrategy::Type::DCT16X32, 0.0f, 0.94f);
+  set(AcStrategy::Type::DCT32X16, 0.0f, 0.94f);
+  set(AcStrategy::Type::DCT32X8, 0.0f, 2.261390410971102f);
+  set(AcStrategy::Type::DCT8X32, 0.0f, 2.261390410971102f);
+  set(AcStrategy::Type::DCT16X8, 0.0f, 0.86f);
+  set(AcStrategy::Type::DCT8X16, 0.0f, 0.86f);
+
+  set(AcStrategy::Type::DCT128X128, 0.0f, 1.0f);
+  set(AcStrategy::Type::DCT128X64, 0.0f, 0.73f);
+  set(AcStrategy::Type::DCT64X128, 0.0f, 0.73f);
+  set(AcStrategy::Type::DCT256X256, 0.0f, 1.0f);
+  set(AcStrategy::Type::DCT256X128, 0.0f, 0.73f);
+  set(AcStrategy::Type::DCT128X256, 0.0f, 0.73f);
+  */
+  struct MergeTry {
+    AcStrategy::Type type;
+    uint8_t priority;
+    float entropy_mul;
+  };
+  const MergeTry kTransformsForMerge[9] = {
+    { AcStrategy::Type::DCT16X8, 2, 0.86f },
+    { AcStrategy::Type::DCT8X16, 2, 0.86f },
+    { AcStrategy::Type::DCT16X16, 3, 0.83f },
+    { AcStrategy::Type::DCT16X32, 4, 0.94f },
+    { AcStrategy::Type::DCT32X16, 4, 0.94f },
+    { AcStrategy::Type::DCT32X32, 5, 0.97f },
+    { AcStrategy::Type::DCT64X32, 6, 1.15f },
+    { AcStrategy::Type::DCT32X64, 6, 1.15f },
+    { AcStrategy::Type::DCT64X64, 8, 1.3f },
+  };
+  uint8_t priority[64] = { 0 };
+  for (int i = 0; i < 9; ++i) {
+    const float entropy_mul = kTransformsForMerge[i].entropy_mul;
+    const uint8_t candidate_priority = kTransformsForMerge[i].priority;
+    AcStrategy acs = AcStrategy::FromRawStrategy(kTransformsForMerge[i].type);
+    for (size_t cy = 0; cy + acs.covered_blocks_y() - 1 < rect.ysize();
+	 cy += acs.covered_blocks_y()) {
+      for (size_t cx = 0; cx + acs.covered_blocks_x() - 1 < rect.xsize();
+	   cx += acs.covered_blocks_x()) {
```

Comment:
What happens if you do `cy++` and `cx++` here instead of doing things naturally aligned? It will surely make things slower, but should also result in better fits (doing things like an 8x16 followed by a 32x16 instead of having to split that 32x16 in an 8x16, a 16x16 and an 8x16).

Reply:
It will be slower, better and it will require a better heuristic than 'priority' to avoid collisions (of for example 16x8 and 8x16 occupying the same 8x8 square).